src/GitVersionCore/MergeMessage.cs

```diff
@@ -4,56 +4,74 @@
-            new MergeMessagePattern("TfsMergeMessageEnglishUS", @"^Merge (?<SourceBranch>[^\s]*) to (?<TargetBranch>[^\s]*)"),
-            new MergeMessagePattern("TfsMergeMessageGermanDE",@"^Zusammengeführter PR ""(?<PullRequestNumber>\d+)""\: (?<SourceBranch>.*) mit (?<TargetBranch>.*) mergen")
+            Pattern("Default", @"^Merge (branch|tag) '(?<SourceBranch>[^']*)'(?: into (?<TargetBranch>[^\s]*))*"),
+            Pattern("SmartGit",  @"^Finish (?<SourceBranch>[^\s]*)(?: into (?<TargetBranch>[^\s]*))*"),
+            Pattern("BitBucketPull", @"^Merge pull request #(?<PullRequestNumber>\d+) (from|in) (?<Source>.*) from (?<SourceBranch>[^\s]*) to (?<TargetBranch>[^\s]*)"),
+            Pattern("GitHubPull", @"^Merge pull request #(?<PullRequestNumber>\d+) (from|in) (?:(?<SourceBranch>[^\s]*))(?: into (?<TargetBranch>[^\s]*))*"),
+            Pattern("RemoteTracking", @"^Merge remote-tracking branch '(?<SourceBranch>[^\s]*)'(?: into (?<TargetBranch>[^\s]*))*"),
+            Pattern("TfsMergeMessageEnglishUS", @"^Merge (?<SourceBranch>[^\s]*) to (?<TargetBranch>[^\s]*)"),
+            Pattern("TfsMergeMessageGermanDE",@"^Zusammengeführter PR ""(?<PullRequestNumber>\d+)""\: (?<SourceBranch>.*) mit (?<TargetBranch>.*) mergen")
         };
 
         public MergeMessage(string mergeMessage, Config config)
         {
             if (mergeMessage == null)
                 throw new NullReferenceException();
 
-            foreach (var pattern in Patterns)
+            foreach(var entry in config.MergeMessageFormats)
             {
-                var match = pattern.Format.Match(mergeMessage);
-                if (match.Success)
+                var pattern = Pattern(entry.Key, entry.Value);
+                if (ApplyPattern(mergeMessage, config.TagPrefix, pattern))
                 {
-                    MatchDefinition = pattern.Name;
-                    MergedBranch = match.Groups["SourceBranch"].Value;
-
-                    if (match.Groups["TargetBranch"].Success)
-                    {
-                        TargetBranch = match.Groups["TargetBranch"].Value;
-                    }
-
-                    if (int.TryParse(match.Groups["PullRequestNumber"].Value, out var pullNumber))
-                    {
-                        PullRequestNumber = pullNumber;
-                    }
-
-                    Version = ParseVersion(MergedBranch, config.TagPrefix);
+                    return;
+                }
+            }
 
-                    break;
+            foreach (var pattern in DefaultPatterns)
+            {
+                if (ApplyPattern(mergeMessage, config.TagPrefix, pattern))
+                {
+                    return;
                 }
             }
         }
 
-        public string MatchDefinition { get; }
-        public string TargetBranch { get; }
-        public string MergedBranch { get; } = "";
+        public string MatchDefinition { get; private set; }
```

Comment:
Is it possible to leave these properties immutable and instead return a new instance of the `MergeMessage` class when mutation is required?

Reply:
These properties are now set in `ApplyPattern` outside of the constructor so required private set access.  I could refactor things so that they are set in the constructor again if you wish (e.g. making `ApplyPattern` a local function and/or merging the pattern collections together).

They are properties of the instance being actively constructed. An internal context class could be used for immutability purposes, but any previous mutations would be redundant as they caller of  would never receive them.

Reply:
@asbjornu This is the only thing left on this review.  Let me know what you want to do here.

Reply:
It's not very important, but if it's not much work to retain immutability, I'd prefer it.